src/device/v4l2/stateless/request.rs

```diff
@@ -0,0 +1,206 @@
+        let mut ctrl: SafeExtControl<T> = ctrl.into();
+        ioctl::s_ext_ctrls(&self.device, which, &mut ctrl).expect("Failed to set output control");
+        self
+    }
+    fn write(&mut self, data: &[u8]) -> &mut Self {
+        self.buffer.write(data);
+        self
+    }
+    fn submit(self) -> PendingRequestHandle {
+        self.buffer.submit(self.timestamp, self.handle.as_raw_fd());
+        self.handle.queue().expect("Failed to queue request handle");
+        PendingRequestHandle {
+            device: self.device.clone(),
+            timestamp: self.timestamp,
+        }
+    }
+}
+
+struct PendingRequestHandle {
+    device: V4l2Device,
+    timestamp: u64,
+}
+
+impl PendingRequestHandle {
+    fn sync(self) -> DoneRequestHandle {
+        DoneRequestHandle {
+            buffer: Rc::new(RefCell::new(self.device.sync(self.timestamp))),
+        }
+    }
+}
+
+struct DoneRequestHandle {
+    buffer: Rc<RefCell<V4l2CaptureBuffer>>,
+}
+
+impl DoneRequestHandle {
+    fn result(&self) -> V4l2Result {
+        V4l2Result {
+            buffer: self.buffer.clone(),
+        }
+    }
+}
+
+#[derive(Default)]
+enum RequestHandle {
+    Init(InitRequestHandle),
+    Pending(PendingRequestHandle),
+    Done(DoneRequestHandle),
+    #[default]
+    Unknown,
+}
+
+impl RequestHandle {
+    fn new(
+        device: V4l2Device,
+        timestamp: u64,
+        handle: ioctl::Request,
+        buffer: V4l2OutputBuffer,
+    ) -> Self {
+        Self::Init(InitRequestHandle::new(device, timestamp, handle, buffer))
+    }
+    fn timestamp(&self) -> u64 {
+        match self {
+            Self::Init(handle) => handle.timestamp,
+            Self::Pending(handle) => handle.timestamp,
+            Self::Done(handle) => handle.buffer.borrow().timestamp(),
+            _ => panic!("ERROR"),
+        }
+    }
+    fn ioctl<C, T>(&mut self, ctrl: C) -> &mut Self
+    where
+        C: Into<SafeExtControl<T>>,
+        T: ExtControlTrait,
+    {
+        match self {
+            Self::Init(handle) => handle.ioctl(ctrl),
+            _ => panic!("ERROR"),
+        };
+        self
+    }
+    fn write(&mut self, data: &[u8]) -> &mut Self {
+        match self {
+            Self::Init(handle) => handle.write(data),
+            _ => panic!("ERROR"),
+        };
+        self
+    }
+    fn submit(self) -> Self {
+        match self {
+            Self::Init(handle) => Self::Pending(handle.submit()),
+            _ => panic!("ERROR"),
+        }
+    }
+    fn sync(self) -> Self {
+        match self {
+            Self::Pending(handle) => Self::Done(handle.sync()),
+            Self::Done(_) => self,
+            _ => panic!("ERROR"),
+        }
+    }
+    fn result(&self) -> V4l2Result {
+        match self {
+            Self::Done(handle) => handle.result(),
+            _ => panic!("ERROR"),
+        }
+    }
+}
+
+pub struct V4l2Request {
+    handle: RefCell<RequestHandle>,
```

Comment:
Having a `RefCell` without a `Rc` usually indicates something is wrong with the mutability of your interface. I think this can be simplified - please see the follow-up CL I've added to your series, and squash it if you agree.